### PR TITLE
Make whitelistening more strict and explicit

### PR DESCRIPTION
**Pull request by take-five**
Previously, whitelistening preprocessor exposed complex fields
structure even they weren't whitelisted. For example, one could
see `{"key": {"foo": "*****", "bar": "*****"}}` in the logs, despite
the fact that `/key` wasn't whitelisted.

This commit makes it so such fields won't be exposed at all, and
example above will produce `{"key":"*****"}`. Also I changed existing
implementation to use Trie data structure, which is much simpler
and (didn't measure) should be faster.

This will require a minor version bump.

### CODE REVIEW

File: spec/preprocessors/whitelist_spec.rb [hunk 1]
```diff
@@ -36,90 +36,96 @@
...
+      )
     end
   end
 
   context 'with whitelisted hash' do
     it 'includes all whitelisted hash elements' do
-      source = {foo: {bar: 'baz'}}
-      target = {foo: {bar: 'baz'}}
+      source = { foo: { bar: 'baz' } }
+      target = { foo: { bar: 'baz' } }
```
Comment (deiwin) [inline spec/preprocessors/whitelist_spec.rb]: Would be nice if formatting changes would be separate.

File: spec/preprocessors/json_pointer_trie.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    it 'returns false if trie does not contain requested prefix or value' do
+      trie.insert('/data/nested/key')
+
+      expect(trie).to_not include('/bad_data')
+      expect(trie).to_not include('/data/bad_nested')
+      expect(trie).to_not include('/data/nested/bad_key')
+    end
+
+    it 'returns true if true contains requested prefix under wildcard' do
```
Comment (deiwin) [inline spec/preprocessors/json_pointer_trie.rb]: `true` -> `trie`

Comment (ostap0207) [inline spec/preprocessors/json_pointer_trie.rb]: `returns true if true`

File: lib/logasm/preprocessors/whitelist.rb [hunk 3]
```diff
@@ -1,28 +1,24 @@
...
   module Preprocessors
     class Whitelist
-
-      DEFAULT_WHITELIST = %w(/id /message /correlation_id /queue)
-      MASK_SYMBOL = '*'
+      DEFAULT_WHITELIST = %w[/id /message /correlation_id /queue].freeze
+      MASK_SYMBOL = '*'.freeze
       MASKED_VALUE = MASK_SYMBOL * 5
-      WILDCARD = '~'
+      WILDCARD = '~'.freeze
```
Comment (ostap0207) [inline lib/logasm/preprocessors/whitelist.rb]: Not used anymore

**Comment by take-five** [comment]
@deiwin
Benchmarks.

Old version:
```
Scalar value whitelisting
                         79.890k (± 4.0%) i/s -    401.316k in   5.031461s
Flat hash whitelisting
                         45.661k (± 6.4%) i/s -    230.877k in   5.079152s
Nested hash whitelisting
                         13.231k (± 3.8%) i/s -     66.350k in   5.022311s
Flat array whitelisting
                         20.840k (± 6.9%) i/s -    105.519k in   5.090336s
```

New version
```
Scalar value whitelisting
                        125.807k (± 6.4%) i/s -    626.496k in   5.004610s
Flat hash whitelisting
                         51.643k (± 5.4%) i/s -    259.794k in   5.046759s
Nested hash whitelisting
                         11.224k (± 3.7%) i/s -     56.628k in   5.052578s
Flat array whitelisting
                         14.449k (±17.4%) i/s -     69.414k in   5.068867s
```

Nested hashes became a bit worse, but not significantly. Also, new implementation won't traverse entire structure, so we shouldn't notice any difference

**Comment by take-five** [comment]
Also, I tried to add cache to Trie implementation and results now are:
```
Calculating -------------------------------------
Scalar value whitelisting
                        234.343k (± 4.8%) i/s -      1.171M in   5.010703s
Flat hash whitelisting
                        108.316k (± 4.1%) i/s -    550.165k in   5.087940s
Nested hash whitelisting
                         25.407k (± 3.4%) i/s -    127.452k in   5.022319s
Flat array whitelisting
                         35.661k (± 4.4%) i/s -    180.284k in   5.066247s
```

File: lib/logasm/preprocessors/json_pointer_trie.rb [hunk 4]
```diff
@@ -17,13 +18,17 @@ def insert(pointer)
       end
 
       def include?(path)
+        @cache.fetch(path) { @cache[path] = traverse_path(path) }
```
Comment (deiwin) [inline lib/logasm/preprocessors/json_pointer_trie.rb]: This looks pretty leaky, depending on the lifecycle of the class instance. I'm assuming it lasts for the entire application.

Comment (take-five) [reply lib/logasm/preprocessors/json_pointer_trie.rb]: Replaced with LRU cache

Comment (take-five) [reply lib/logasm/preprocessors/json_pointer_trie.rb]: New results are
```
Scalar value whitelisting
                        190.786k (± 4.4%) i/s -    958.230k in   5.032744s
Flat hash whitelisting
                         92.370k (± 3.1%) i/s -    468.396k in   5.075706s
Nested hash whitelisting
                         22.587k (± 3.7%) i/s -    114.088k in   5.058340s
Flat array whitelisting
                         30.929k (± 4.4%) i/s -    155.346k in   5.032732s
```

**Comment by indrekj** [comment]
!merge

**Pull request merged**